truffle/src/main/java/org/jruby/truffle/nodes/ext/BigDecimalNodes.java

```diff
@@ -1180,8 +1392,62 @@ public RubyBasicObject absSpecial(RubyBasicObject value) {
 
     }
 
+    @CoreMethod(names = "round", optional = 2)
+    public abstract static class RoundNode extends BigDecimalCoreMethodArrayArgumentsNode {
+
+        public RoundNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+        }
+
+        @TruffleBoundary
+        private BigDecimal round(RubyBasicObject value, int digit, RoundingMode roundingMode) {
+            final BigDecimal valueBigDecimal = getBigDecimalValue(value);
+            final int end = -valueBigDecimal.scale();
+            final int start = valueBigDecimal.precision() + end;
+            final int newPrecision = start - digit;
+            return valueBigDecimal.round(new MathContext(newPrecision, roundingMode));
+        }
+
+        @Specialization(guards = "isNormal(value)")
+        public RubyBasicObject round(VirtualFrame frame, RubyBasicObject value, NotProvided digit, NotProvided roundingMode) {
+            return createBigDecimal(frame, round(value, 0, getRoundMode(frame)));
+        }
+
+        @Specialization(guards = "isNormal(value)")
+        public RubyBasicObject round(VirtualFrame frame, RubyBasicObject value, int digit, NotProvided roundingMode) {
+            return createBigDecimal(frame, round(value, digit, getRoundMode(frame)));
+        }
+
+        @Specialization(guards = "isNormal(value)")
+        public RubyBasicObject round(VirtualFrame frame, RubyBasicObject value, int digit, int roundingMode) {
+            return createBigDecimal(frame, round(value, digit, toRoundingMode(roundingMode)));
+        }
+
+        @Specialization(guards = "!isNormal(value)")
+        public RubyBasicObject roundSpecial(VirtualFrame frame, RubyBasicObject value, Object precision, Object roundingMode) {
+            switch (getBigDecimalType(value)) {
+                case NEGATIVE_INFINITY:
+                    CompilerDirectives.transferToInterpreter();
+                    throw new RaiseException(getContext().getCoreLibrary().
+                            floatDomainError("Computation results to '-Infinity'", this));
+                case POSITIVE_INFINITY:
+                    CompilerDirectives.transferToInterpreter();
+                    throw new RaiseException(getContext().getCoreLibrary().
+                            floatDomainError("Computation results to 'Infinity'", this));
+                case NEGATIVE_ZERO:
+                    return value;
+                case NAN:
+                    CompilerDirectives.transferToInterpreter();
+                    throw new RaiseException(getContext().getCoreLibrary().
+                            floatDomainError("Computation results to 'NaN'(Not a Number)", this));
+            }
+            CompilerAsserts.neverPartOfCompilation();
```

Comment:
I would like to see a message in the `UnsupporteOperationException` to help with debugging.  But the `neverPartOfCompilation` is the part I was questioning.  I didn't know it could float in a method like that.